Problem 1
Part (a)

In [21]:
import numpy as np

appList=[]
def f(x):
    return np.exp(x)
for i in range(10):
    h=0.1/2**i
    points=[1/2-h,1/2,1/2+h]
    appList.append((f(points[2])-f(points[0]))/(2*h))
    print(appList[i])

print(appList - f(1/2))

1.6514705137461927
1.6494083237722634
1.648893017866171
1.6487642064853247
1.6487320045835219
1.6487239541670817
1.6487219415666488
1.6487214384167714
1.648721312629391
1.6487212811824747
[2.74924305e-03 6.87053072e-04 1.71747166e-04 4.29357852e-05
 1.07338834e-05 2.68346695e-06 6.70866521e-07 1.67716643e-07
 4.19292627e-08 1.04823465e-08]


As h is halved, the error is roughly 1/4=(1/2)^2 of before. Hence the rate of convergence is indeed h^2.

Part (b)

In [22]:
appList1=[]
for i in range(40):
    h=0.1/2**i
    points=[1/2-h,1/2,1/2+h]
    appList1.append((f(points[2])-f(points[0]))/(2*h))
print(appList1 - f(1/2))

[ 2.74924305e-03  6.87053072e-04  1.71747166e-04  4.29357852e-05
  1.07338834e-05  2.68346695e-06  6.70866521e-07  1.67716643e-07
  4.19292627e-08  1.04823465e-08  2.62147015e-09  6.55824728e-10
  1.62423852e-10  3.96420674e-11  3.26227934e-12 -1.49276147e-11
 -1.49276147e-11 -8.76871908e-11 -8.76871908e-11 -8.76871908e-11
 -6.69763800e-10  4.94389418e-10  5.15100229e-09 -4.16222345e-09
 -4.16222345e-09 -2.27886749e-08  5.17171310e-08 -2.27886749e-08
 -1.71800287e-07  4.24246161e-07  4.24246161e-07 -1.95993963e-06
  4.24246161e-07  9.96098932e-06  9.96098932e-06 -9.11249700e-06
 -9.11249700e-06  1.43475394e-04  1.43475394e-04 -1.61700388e-04]


Based on the error approximation, h=0.1/(2)^14 corresponding to an error of 3.26227934*10^(-12) appears to be the optimal h.

Part (c)

By extrapolation, we may construct $$D^*_h(f')=\frac{4D_{h/2}(f')-D_h(f')}{3}$$\
which will have an error of $O(h^4)$.

In [32]:
appList2 = []
for i in range(20):
    h=0.1/2**(i+1)
    points=[1/2-h,1/2,1/2+h]
    appList2.append((f(points[2])-f(points[0]))/(2*h))
    
appExtrap = []
for i in range(20):
    appExtrap.append((4*appList2[i]-appList1[i])/3)
    print("D*_h where h=0.1/2^" + str(i) + " is " + str(appExtrap[i]))
    print("The Error for this h is " + str(f(1/2)-appExtrap[i]))

D*_h where h=0.1/2^0 is 1.6487209271142869
The Error for this h is 3.43585841333649e-07
D*_h where h=0.1/2^1 is 1.6487212492308068
The Error for this h is 2.1469321431766275e-08
D*_h where h=0.1/2^2 is 1.648721269358376
The Error for this h is 1.3417522648495606e-09
D*_h where h=0.1/2^3 is 1.6487212706162542
The Error for this h is 8.38740188413567e-11
D*_h where h=0.1/2^4 is 1.648721270694935
The Error for this h is 5.193179219986632e-12
D*_h where h=0.1/2^5 is 1.6487212706998378
The Error for this h is 2.9043434324194095e-13
D*_h where h=0.1/2^6 is 1.6487212707001457
The Error for this h is -1.7541523789077473e-14
D*_h where h=0.1/2^7 is 1.648721270700264
The Error for this h is -1.3589129821411916e-13
D*_h where h=0.1/2^8 is 1.6487212707001693
The Error for this h is -4.107825191113079e-14
D*_h where h=0.1/2^9 is 1.6487212707013061
The Error for this h is -1.177946629127291e-12
D*_h where h=0.1/2^10 is 1.6487212707007377
The Error for this h is -6.095124405192109e-13
D*_h where h=0.

Every error is roughly 1/16 of the previous one as h is halved, which confirms the convergence rate of O(h^4).
According to the error approximations, the Optimal h appears to be 0.1/2^6. 